feat: dark mode

mtfoley opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/open-sauced/blob/HEAD/CONTRIBUTING.md#create-a-pull-request.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/open-sauced/blob/HEAD/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

## What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description
PR to fix #607 by adding dark mode support.  Tested a handful of things so far in chrome, but plenty of work to do.  Will try to get out of draft next week if all looks in the right direction.

<!-- Please do not leave this blank -->

## Related Tickets & Documents
<!-- Please use this format link issue numbers: Fixes #123 https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword-->


## Mobile & Desktop Screenshots/Recordings (if there are UI changes)



## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 readme
- [ ] 📜 contributing.md
- [ ] 📓 docs
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?
Make sure it looks nice 👍 


## [optional] What gif best describes this PR or how it makes you feel?
![cruella](https://media.giphy.com/media/9ayXNVkco6fM6YUNPU/source.gif)


<!-- note: PRs with deletes sections will be marked invalid -->

mtfoley commented:
@bdougie @filiptronicek @0-vortex I've got this draft PR started and I definitely have some more tweaks to make - icons, some muted text, etc. please take a look at the Netlify deploy previews and let me know what y'all feel would drive to this being ready to pull out of draft.

filiptronicek reviewed (commented):
I'm really glad this is being worked on. I like the changes so far, maybe I would make the `#0C0E0E` which is used for many elements a lighter grey, but it's looking good.

We could also add a user preference switch for people who prefer either design independent of their set system theme. Like so:
![image](https://user-images.githubusercontent.com/29888641/120335445-c3093700-c2f1-11eb-83a4-6b4ae9e46a31.png)

mtfoley commented:
@filiptronicek thx for taking a look. Most of my changes so far have been adjusting adding CSS with a media selector. In order give the user some kind of on demand selection, I think I would need to backtrack these to use CSS that targets based on the body element class and have JS that adjusts that class, instead of using media selector... does that sound right?

filiptronicek commented:
> @filiptronicek thx for taking a look. Most of my changes so far have been adjusting adding CSS with a media selector. In order give the user some kind of on demand selection, I think I would need to backtrack these to use CSS that targets based on the body element class and have JS that adjusts that class, instead of using media selector... does that sound right?

Yes, that's about how I feel like it can be done. We could make it so the js adds a class when the page loads and also adds an event listener to change the class even when the user changes the color scheme preference when the page is already loaded.

mtfoley commented:
Ok, I don't have any past experience with the styled-components but I'll see what it lets me do :-) any prior used components in the project that you'd suggest for mode choosing, and where to place it?

filiptronicek commented:
> 
> 
> Ok, I don't have any past experience with the styled-components but I'll see what it lets me do :-) any prior used components in the project that you'd suggest for mode choosing, and where to place it?

@mtfoley although this modal was not done in React, here is the minimal version of this, which saves the preference to the local storage: https://codepen.io/filip_tronicek/pen/RwpxqKo

bdougie commented:
Sorry that comment may be pre-mmature. I am interested in this `prefers-color-scheme: dark`, I have not used or heard of that.

mtfoley commented:
In order to test it I had to go into Chrome Developer Tools / Rendering, and I think the setting is "Emulate CSS prefers-color-scheme: dark" or something to that effect. On my iPhone the deploy preview shows up in dark mode because that's what I have picked in display settings. On my MacBook it changes with the time of day, I think because of the dynamic desktop photo thing.

bdougie commented:
> In order to test it I had to go into Chrome Developer Tools / Rendering, and I think the setting is "Emulate CSS prefers-color-scheme: dark" or something to that effect. On my iPhone the deploy preview shows up in dark mode because that's what I have picked in display settings. On my MacBook it changes with the time of day, I think because of the dynamic desktop photo thing.

This might need some documentation included with the PR. I was not familiar with all that myself.

0-vortex commented:
## Some screenshots from Big Sur 11.4 Beta (20F5055c)

### Safari 14.1.1 (16611.2.5):

<img width="1680" alt="Screenshot 2021-06-01 at 21 13 43" src="https://user-images.githubusercontent.com/237133/120378160-5b1d1580-c31e-11eb-9255-58e1f1d90422.png">

### Chrome 92.0.4515.20 (Official Build) dev (x86_64):

<img width="1680" alt="Screenshot 2021-06-01 at 21 15 54" src="https://user-images.githubusercontent.com/237133/120378432-b3ecae00-c31e-11eb-8385-786d322dcee3.png">

The color scheme seems all good without checking any experiments, the only thing I am noticing is different font x-width and unrelated to this PR, the use-shopping-cart could use some re-thinking.

mtfoley commented:
Ok, after reading all the commentary I think I'm going to take a step back and sort of start over with this. I think I have some reading to do on this article: https://dev.to/nw/adding-dark-mode-to-your-react-app-with-hooks-media-queries-and-css-variables-50h0

To me, the behavior scope sounds like:
- app detect user's current `prefers-color-scheme` value, as well as any overrides in local storage.
- app adjust the body or html element class to reflect the active color scheme and let CSS take over from that point

And the implementation scope sounds like it needs to include some documentation as well for this.

I'm not sure how this approach will work with the styled-components, but we'll find out soon enough.

0-vortex commented:
> Ok, after reading all the commentary I think I'm going to take a step back and sort of start over with this. I think I have some reading to do on this article: [dev.to/nw/adding-dark-mode-to-your-react-app-with-hooks-media-queries-and-css-variables-50h0](https://dev.to/nw/adding-dark-mode-to-your-react-app-with-hooks-media-queries-and-css-variables-50h0)
> 
> To me, the behavior scope sounds like:
> 
> * app detect user's current `prefers-color-scheme` value, as well as any overrides in local storage.
> * app adjust the body or html element class to reflect the active color scheme and let CSS take over from that point
> 
> And the implementation scope sounds like it needs to include some documentation as well for this.
> 
> I'm not sure how this approach will work with the styled-components, but we'll find out soon enough.

I feel like that article is needlessly complicated, trying to cover space instead of simplifying what it's doing and instead go for a "tiered" learning approach:
- styled-components docs: https://styled-components.com/docs/advanced#via-usecontext-react-hook
- dark mode mental framework: https://css-tricks.com/dark-modes-with-css/
- styled-components demo: https://codesandbox.io/s/3ljgt?file=/src/index.js 
- building a react-dark-toggle with styled-components: https://www.smashingmagazine.com/2020/04/dark-mode-react-apps-styled-components/ - would replace your linked tutorial with this one
- and a common used package for anything but the styling itself: https://www.npmjs.com/package/react-toggle-dark-mode

Not being too into styled-components I would split 3 ways, toggle component, set up `<ThemeProvider>` to check local storage for `ThemeMode ` value and create a function to manipulate `ThemeMode`. Then I would able to use that 2 way switch anywhere by just reading the styled-components theme from the react hooks. This is how that split would roughly look like, this could be `Header` component leveraging `react-toggle-dark-mode`:

```jsx
import { ThemeContext } from 'styled-components';
const themeContext = useContext(ThemeContext);

<DarkModeSwitch
	  checked={themeContext === 'light'}
	  onChange={() => {
	    setDarkMode(themeContext === 'light' ? 'light' : 'dark');
	  }}
	  size={42}
/>
```

Hope this removes some confusion, learning 3 things at once you already got 2 of them right 🤩

mtfoley commented:
Well that was fun - I feel like I'm getting close now.
I've got a ThemeContext and I'm differentiating between dark/light/system, falling back to system.
I've got an unstyled triple button that lets you update the theme when logged in.
I changed my media queries to use `body.class & { }` instead.

Known issues:
- the buttons need to look nice. Maybe have icons like in some examples I've seen. Not sure what icon to use for "system" or "automatic".
- the buttons don't adjust the theme when not logged in... need to take some time with that (also, should we even show them?)
- I think I've still clobbered the background of the search field in AddRepoForm. I need to look back at the original style of it in main branch.
- I need to add docs.

notBdougie commented:
Perhaps this HEX instead of black #241E1D

bdougie commented:
I switched my phone to dark mode to get a feel fork dark mode in different apps and like twitters lines in between tweets. I think what you have so far is really good. 

I don't think you really need to have it. Lose to the original hue at all. Disregard my comment on stream. 

The buttons to switch it on and off is a bonus.

![image](https://user-images.githubusercontent.com/5713670/120883599-e1419200-c592-11eb-94ba-989bb0e7f314.png)

mtfoley reviewed (commented):
don't know how to turn this review off...

mtfoley commented:
@bdougie @0-vortex @filiptronicek I think we are within striking distance. 

As a bonus, this PR deletes that typo where the word "repo" ended up next to the repo language labels :-)

filiptronicek reviewed (changes requested):
This PR is indeed very close to perfection, I just have a couple things:

- The icons may look a lot better if they were white in all theme settings (we can just use [`filter: invert()`](https://developer.mozilla.org/en-US/docs/Web/CSS/filter-function/invert()) for this.)
- The skeleton loading is still in light colors, and that looks a bit off
- Maybe add a bit of a margin to the theme selection? 

![image](https://user-images.githubusercontent.com/29888641/120901782-20c4b880-c63d-11eb-9ca6-c95a486043b8.png)
![image](https://user-images.githubusercontent.com/29888641/120901862-8dd84e00-c63d-11eb-9afe-f6d50c1aeb75.png)

mtfoley commented:
Alright I'll take a look at these. The octicons render svg straight to the DOM so I think those just inherit fill by the font color. 

For the skeleton stuff I think I may need to play with className attributes since it's been mostly CSS approach so far and for the most part the theme context doesn't have a lot of effect on the components in the tree. I think the more neutral grey used in some of these can probably stay as is, it's the card background that needs to match the hydrated state.

bdougie commented:
I personally think the svgs can stay as is until a solution is discovered. I think we can open a good first issue for that honestly. I can take a closer look at everything when I am back at a keyboard.
<details><summary>Browser metadata</summary>





```
Path:      /
Browser:   Mobile Safari 14.1.1 on iOS 14.6
Viewport:  375 x 635 @3x
Language:  en-us
Cookies:   Disabled
```

[View on BrowserStack](https://live.browserstack.com/dashboard#url=https%3A%2F%2Fdeploy-preview-1020--open-sauced.netlify.app%2F&os=iOS&os_version=14.6&speed=1&start=true)



</details>

[Open Deploy Preview](https://deploy-preview-1020--open-sauced.netlify.app/) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60bbbd7cf5c61a00070fda70&commentID=60bbddcb6deb950b376ad519&resolution=resolved)

0-vortex commented:
I have experience using the filter function on monochrome icons, as @filiptronicek mentioned you can use invert to create any color even without the fill on svg. Have some accompanying code for that in my cf-workers repo if needed 🍕

mtfoley commented:
I use filter: invert(0.85) in index.css with the current changes, and I prob just have more find & replace to run for the remainder of the relevant img tags. At one point I was inverting *all* img tags and the splash screen gave me a good chuckle. But yeah, never had to do anything with it before :-)

mtfoley commented:
And I forgot docs...

mtfoley commented:
@filiptronicek @bdougie @0-vortex  I made the updates you asked for here.  

- [x]  For [react-loading-skeleton](https://github.com/dvtng/react-loading-skeleton/blob/master/src/skeleton.js#L18), I added to index.css to handle the color variables in dark mode.  
- [x] From what I found in searching for img tags, I had just the new recommended repos section to add the svg class to.  
- [x] Docs: I added a section toward the end of README.md.

bdougie commented on src/components/Repository.js:
```diff
@@ -211,7 +211,7 @@ function Repository({user, match}) {
               </div>
             </span>
           ) : (
-            <SkeletonTheme color="#fbfbfb" highlightColor="#f2f2f2">
```
Thanks for moving this

bdougie reviewed (approved):
This looks good to me. I don't have a lot of context on approaching dark mode technically, but all this makes sense to me. 

My only ask would be adding links to blog posts or guides you used to solve this to the PR body. Mainly because, future contributors will have questions on the decisions made and I want to point to this PR in the future.

mtfoley commented:
@bdougie I don't think I caught what you asked me to add in that first sentence, other than links to what I referred to.

RE: The general methodology, I couldn't find any exact precedent because some tutorials are written around using prefers-color-scheme to determine an appropriate theme, and others are written around sort of controlling the theme completely within the app in a two-state paradigm, but I didn't see any that gave examples of how to do both (three states like what we have here). Open Sauced might be somewhat unique in that after the PR (maybe?) The rationale of falling back to system preferences was largely based on the StackOverflow [blog](https://stackoverflow.blog/2020/03/31/building-dark-mode-on-stack-overflow/) that @0-vortex pointed out.  I didn't wind up using any of the toggle buttons
RE: img tags with src={svg}, I referred to this [post](https://medium.com/@union_io/swapping-fill-color-on-image-tag-svgs-using-css-filters-fa4818bf7ec6) and [this](https://catalin.red/svg-favicon-light-dark-theme/).
RE: ThemeContext. I used a context because I wanted to define the theme as high up in the component tree as I could, but I wasn't sure how deeply in the component tree I would need to interact with it. [React Context Docs](https://reactjs.org/docs/context.html#consuming-multiple-contexts)

However, even though I use a context, I actually refer to the context very little if at all other than the toggle buttons themselves.  This is because the "system" state of the context is still somewhat indeterminate from any components' perspective of choosing between light and dark.  Because of this, no components actually choose between light and dark - the CSS file and the CSS portion of the styled-components code leverage the `<body class="dark">` to decide how to look.

filiptronicek reviewed (changes requested):
Everything looks good, just my background color does not work properly on Firefox...

**Brave**:
![image](https://user-images.githubusercontent.com/29888641/121414150-ac548700-c966-11eb-95f7-9f467382a71a.png)

**Firefox Dev**:
![image](https://user-images.githubusercontent.com/29888641/121414335-cf7f3680-c966-11eb-9403-9601e6ac990b.png)


Also, can we make the search icon also inverted? Would be a nice touch.
![image](https://user-images.githubusercontent.com/29888641/121414442-ed4c9b80-c966-11eb-9bfe-260fc08f1535.png)

mtfoley commented:
I think I've got firefox fixed now, and I ended up just creating a 2nd version of the search background SVG for dark mode. Still just using body.dark class to differentiate. 

I also made a whole bunch more tweaks to things like the issues icons, form field colors, and rendered note colors.

bdougie commented:
can confirm that firefox looks better. going to merge. If anything else is caught, we can move it to an issue. Thanks a ton for taking this on @mtfoley and thanks for testing other browsers @filiptronicek. I got to get out of my chrome for everyone's mindset.  

<img width="1552" alt="Screen Shot 2021-06-09 at 7 04 03 PM" src="https://user-images.githubusercontent.com/5713670/121453164-8b376a80-c955-11eb-9bac-0dac1202e5f8.png">
